File: 프로그래머스/2/152996. 시소 짝꿍/시소 짝꿍.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool Comp(int a, int b) {
    return a > b;
}

bool Seesaw(int a, int b) {
    if(a == b)
        return true;
    if(a == 2 * b)
        return true;
    if(2 * a == 3 * b)
        return true;
    if(3 * a == 4 * b)
        return true;
    
    return false;
}

long long solution(vector<int> weights) {
    long long answer = 0;
    
    sort(weights.begin(), weights.end(), Comp);
    
    for(int i = 0; i < weights.size() - 1; i++) {
        for(int j = i + 1; j < weights.size(); j++) {
            if(Seesaw(weights[i], weights[j]))
                answer++;
        }
    }
    
    return answer;
}

File: 프로그래머스/2/152996. 시소 짝꿍/시소 짝꿍.js
function solution(weights) {
    let answer = 0;
    
    const cnt = Array.from({length: 1001}, () => 0);
    
    // 몸무게 인원 체크
    for(let i = 0; i < weights.length; i++)
        cnt[weights[i]]++;
    
    // 2:3, 2:4(1:2), 3:4
    for(let i = 0; i < weights.length; i++){
        if(weights[i] % 2 === 0){
            let check = (weights[i] / 2) * 3;
            if(check <= 1000)
                answer += cnt[check];
        }
        
        if(weights[i] % 3 === 0){
            let check = (weights[i] / 3) * 4;
            if(check <= 1000)
                answer += cnt[check];
        }
        
        let check = weights[i] * 2;
        if(check <= 1000)
            answer += cnt[check];
    }
    
    for(let i = 100; i <= 1000; i++){
        if(cnt[i] >= 2)
            answer += cnt[i] * (cnt[i] - 1) / 2;
    }
    
    return answer;
}